diesel-factories-code-gen/src/lib.rs

```diff
@@ -1,237 +1,368 @@
+    }
+
+    fn table_path(&self) -> &syn::Path {
+        &self.options.table
+    }
+
+    fn factory_generics(&self) -> &syn::Generics {
+        &self.input.generics
+    }
+
+    fn struct_fields(&self) -> syn::punctuated::Iter<syn::Field> {
+        use syn::{Data, Fields};
+
+        match &self.input.data {
+            Data::Union(_) => panic!("Factory can only be derived on structs"),
+            Data::Enum(_) => panic!("Factory can only be derived on structs"),
+            Data::Struct(data) => match &data.fields {
+                Fields::Named(named) => named.named.iter(),
+                Fields::Unit => panic!("Factory can only be derived on structs with named fields"),
+                Fields::Unnamed(_) => {
+                    panic!("Factory can only be derived on structs with named fields")
+                }
+            },
+        }
+    }
+
+    fn diesel_insert_values(&self) -> Vec<TokenStream> {
+        self.struct_fields()
+            .map(|field| self.diesel_insert_value(field))
+            .collect()
+    }
+
+    fn diesel_insert_value(&self, field: &syn::Field) -> TokenStream {
+        let name = field
+            .ident
+            .as_ref()
+            .unwrap_or_else(|| panic!("Factory can only be derived for named fields"));
+
+        if let Some(association) = self.parse_association_type(&field.ty) {
+            let foreign_key_field = ident(&format!("{}_id", name));
+            if association.is_option {
+                quote! {
+                    {
+                        let value = self.#name.map(|inner| {
+                            inner.insert_returning_id(con)
+                        });
+                        #foreign_key_field.eq(value)
+                    }
+                }
             } else {
-                None
+                quote! {
+                    #foreign_key_field.eq(self.#name.insert_returning_id(con))
+                }
             }
-        })
-        .collect::<Vec<_>>();
-
-    let association_field_impls = fields
-        .iter()
-        .filter_map(|field| {
-            let name = field.ident.as_ref().unwrap_or_else(|| panic!("Field without name"));
-            let ty = &field.ty;
-            let type_string = ty.into_token_stream().to_string();
-            let re = Regex::new(r"^Option < .* >$").unwrap();
-            let options = re.captures(&type_string);
-            let optional = options.is_some();
-            let maybe_factory_attr = field.attrs.iter().find(|attr| {
-                attr.path
-                    .segments
-                    .iter()
-                    .any(|segment| &segment.ident.to_string() == "factory")
+        } else {
+            quote! {
+                #name.eq(&self.#name)
+            }
+        }
+    }
+
+    fn is_association_field(&self, ty: &syn::Type) -> bool {
+        let as_string = self.type_to_string(ty);
+        as_string.contains("Association <")
```

Comment:
I think we should check that it _starts_ with ```Association```. It's a bit of an edge case, but contains will think ```MyCrateAssociation <``` is one of our Association types...

Reply:
It also needs to supports `Option<Association<M, F>>`, but yeah I would like to find a nicer solution. Maybe `contains(" Association <")` will work since token streams get spaces between each token.

Sorry for merging this before letting you review by the way, I was just really excited to tell Reddit :P

Reply:
All good, I'm a huge fan of merge early, merge often :shipit: